### DISCUSSION TITLE
Changing targets and creating sequences

### DISCUSSION THREAD

**Discussion by darcangelomauro** [pasqal-io/Pulser#139] [Ideas]
Hi everybody, I'm a freshly hired employee at Pasqal.

I was going through the tutorials in the arxiv paper and there was something I found confusing about adding pulses to a local channel in a sequence.

Basically what I found myself doing was to use Sequence.target every time before a Sequence.add, which made me think that perhaps it would be possible to pass the target as an optional argument to Sequence.add?

Also, even when I knew the target was already the correct one, I would specify it anyway before Sequence.add, otherwise I'm sure I would lose track eventually of what I was doing. So I ended up with a lot of warnings:
UserWarning: The provided qubits are already the target. Skipping this target instruction.

I'm not sure about possible hardware constraints so I thought it would be useful to post this comment here.

Mauro

**Comment by HGSilveri** [comment]
Hi Mauro, welcome to the team!

I see what you mean. The reason behind that is that, in some cases, the `target` instruction can consume some time, so it is important not to do it unnecessarily and to be mindful of when you're doing it. If you're aware of when you're changing your target, and you know that when you do it you might waste some it, you would naturally tend to group all operations on the same target as much as possible.

I have to say I like your idea of having the "target" as an argument for the `add` command, but I'm afraid of the lightness in usage it might provoke in a user, which would then just start retargeting a channel unnecessarily.

That being said, it is up for debate on whether that `UserWarning` can be suppressed. I agree it's not particularly useful, but it's not harmful either, it's just a warning after all. It all comes down as to whether we want to discourage what you're currently doing or not.

**Comment by darcangelomauro** [comment]
Hi,

I see, so if the target argument is added to `add`, you are afraid users might start to be less mindful about optimizing their retargeting operations?
An interesting project then would be to build an optimizer that reconfigures a sequence in such a way as to minimize retargeting or other time-consuming parts (unless this is already a thing, I'm not sure).

**Reply by HGSilveri** [reply]
> I see, so if the target argument is added to `add`, you are afraid users might start to be less mindful about optimizing their retargeting operations?

Yes, that's it.

> An interesting project then would be to build an optimizer that reconfigures a sequence in such a way as to minimize retargeting or other time-consuming parts (unless this is already a thing, I'm not sure).

I'm not sure about that at the pulse level, but it might make sense to implement at the quantum circuit level.
